extract-hoogle-hackage-info/src/Main.hs

```diff
@@ -0,0 +1,305 @@
+          Nothing -> HashMap.insert packageName' (HashMap.insert infixOpName [] packageFixityMap) sPackageToOps
+          Just _ -> sPackageToOps
+      infixOpName = declToInfixName . unpack $ declOpName
+      packageName' = unpack packageName
+   in
+    state{sPackageToOps = sPackageToOps'}
+
+onFixityDecl :: Text -> State -> (Text, Text, Text) -> State
+onFixityDecl packageName state@State{..} (rawFixDir, rawFixPrec, infixOpName) =
+  let (sPackageToOps', sConflicts') = case HashMap.lookup packageName' sPackageToOps of
+        Nothing | packageName' == "base" -> (HashMap.insert packageName' (HashMap.insert infixOpName' [fixDecl] baseInitialFixityMap) sPackageToOps, sConflicts)
+        Nothing -> (HashMap.insert packageName' (HashMap.singleton infixOpName' [fixDecl]) sPackageToOps, sConflicts)
+        Just packageFixityMap -> case fromMaybe [] $ HashMap.lookup infixOpName' packageFixityMap of
+          [] -> (HashMap.insert packageName' (HashMap.insert infixOpName' [fixDecl] packageFixityMap) sPackageToOps, sConflicts)
+          fixDecls | fixDecl `elem` fixDecls -> (sPackageToOps, sConflicts)
+          fixDecls -> (HashMap.insert packageName' (HashMap.insert infixOpName' (fixDecl : fixDecls) packageFixityMap) sPackageToOps, HashMap.insert (packageName', infixOpName') (fixDecl : fixDecls) sConflicts)
+      packageName' = unpack packageName
+      infixOpName' = unpack infixOpName
+      fixDecl = let fixPrec = read . unpack $ rawFixPrec in FixityInfo {fixDir = Just (readFixDir . unpack $ rawFixDir), fixMinPrec = fixPrec, fixMaxPrec = fixPrec}
+      readFixDir = \case
+        "infix" -> InfixN
+        "infixr" -> InfixR
+        "infixl" -> InfixL
+        other -> error $ "unexpected fixity direction: " ++ other
+   in
+    state{sPackageToOps = sPackageToOps', sConflicts = sConflicts'}
+
+finalizePackageToOps :: HashMap String (HashMap String [FixityInfo]) -> HashMap String FixityMap
+finalizePackageToOps = HashMap.map (HashMap.map finalize) where
+  finalize = \case
+    [] -> unspecifiedFixityInfo
+    fs -> sconcat . NE.fromList $ fs
+
+extractFixitiesFromFile ::
+  -- | Hoogle database path
+  FilePath ->
+  -- | Current file path
+  FilePath ->
+  StateT State IO ()
+extractFixitiesFromFile hoogleDatabasePath filePath = do
+  liftIO $ putFmtLn "Processing file {}..." (Only filePath)
+  state@State{sProcessedFiles} <- get
+  fileContent <- liftIO . readFileUtf8Latin1 $ filePath
+  packageName <- liftIO . fromRightWithIOError $ getPackageName hoogleDatabasePath filePath
+  let state' = foldl' (onSymbolDecl packageName) state (subgroups1 <$> symbolDecls fileContent)
+      state'' = foldl' (onFixityDecl packageName) state' (subgroups3 <$> fixityDecls fileContent)
+      subgroups1 (_ :| g1 : []) = g1
+      subgroups1 _ = error "invalid number of capturing groups"
+      subgroups3 (_ :| g1 : g2 : g3 :[]) = (g1, g2, g3)
+      subgroups3 _ = error "invalid number of capturing groups"
```

Comment:
You can elide the `error` calls and make this more typesafe by using [type-indexed capture groups](https://hackage.haskell.org/package/pcre2-2.0.3/docs/Text-Regex-Pcre2.html#g:7). See here for the minimal diff: https://gist.github.com/amesgen/90fe533fea1c14cda4f86f70d657259d

Reply:
Done!